src/main/app/src/components/koulutus/ToteutusList.tsx

```diff
@@ -0,0 +1,209 @@
+  oid: string;
+};
+
+type JarjestajaData = {
+  jarjestajat: Jarjestaja[];
+  loading: boolean;
+  sortedFilters: Record<string, FilterType[]>;
+};
+
+const getQueryStr = (values: Array<{ id: string }>) =>
+  values.map(({ id }) => id).join(',');
+
+export const ToteutusList = ({ oid }: Props) => {
+  const classes = useStyles();
+  const { jarjestajat, loading, sortedFilters }: JarjestajaData = useSelector(
+    selectJarjestajat
+  );
+
+  const valinnatFromHaku = useSelector(getSuodatinValinnatProps);
+  const initialValues: Record<string, FilterType[]> = useMemo(
+    () =>
+      _fp.pipe(
+        _fp.pick(['opetuskieli', 'sijainti', 'opetustapa']),
+        // TODO: Refactor name to nimi in state
+        _fp.mapValues((arr: any[]) =>
+          arr.map(({ name, ...rest }) => ({ nimi: name, ...rest }))
+        )
+      )(valinnatFromHaku) as any,
+    [valinnatFromHaku]
+  );
+
+  const { t } = useTranslation();
+  const dispatch = useDispatch();
+
+  const getLocalizedMaksullisuus = (isMaksullinen: boolean, maksuAmount: number) =>
+    isMaksullinen ? `${maksuAmount} €` : t('toteutus.maksuton');
+
+  const [chosenFilters, setChosenFilters] = useState(initialValues);
+  const chosenFilterCount = useMemo(
+    () => _fp.sum(Object.values(chosenFilters).map((v) => v.length)),
+    [chosenFilters]
+  );
+
+  const handleFilterChange = useCallback(
+    (newChosenFilters: object) => {
+      const usedFilters = { ...chosenFilters, ...newChosenFilters };
+      setChosenFilters(usedFilters);
+      const queryStrings = _fp.mapValues(getQueryStr, usedFilters);
+      dispatch(fetchKoulutusJarjestajat(oid, queryStrings));
+    },
+    [dispatch, oid, chosenFilters]
+  );
+
+  const handleFiltersClear = useCallback(() => {
+    const usedFilters = _fp.mapValues((_) => [], chosenFilters);
+    setChosenFilters(usedFilters);
+    const queryStrings = _fp.mapValues(getQueryStr, usedFilters);
+    dispatch(fetchKoulutusJarjestajat(oid, queryStrings));
+  }, [dispatch, oid, chosenFilters]);
+
+  // Initial fetch with params from Haku
+  useEffect(() => {
+    const queryStrings = _fp.mapValues(getQueryStr, initialValues);
+    dispatch(fetchKoulutusJarjestajat(oid, queryStrings));
+  }, [dispatch, oid, initialValues]);
+
+  const someSelected = useMemo(
+    () => Object.values(chosenFilters).some((v) => v.length > 0),
+    [chosenFilters]
+  );
+
+  return (
+    <Container maxWidth="lg" className={classes.container}>
+      <Typography variant="h2">{t('koulutus.tarjonta')}</Typography>
+      <Spacer />
+      <Hidden smDown>
+        <Grid
+          container
+          direction="row"
+          justify="center"
+          spacing={2}
+          className={classes.filtersContainer}>
+          <Grid item className={classes.filter}>
+            <OpetuskieliSuodatin
+              handleFilterChange={handleFilterChange}
+              initialValues={initialValues.opetuskieli}
+              sortedValues={sortedFilters.opetuskieli}
```

Comment:
```suggestion
              sortedValues={sortedFilters?.opetuskieli}
```
Voi olla turha, mutta mietityttää että pitäisikö olla aina optional chaining mukana

Reply:
Selectoreissa on defaulttiarvot sisässä joten ei pitäis tulla tilannetta jossa näitä objekteja ei ole.

Optional chainingia pitäs kuitenkin käyttää lähinnä jos halutaan tukea tilannetta jossa jotain ei ole varmalle mukana. Typescriptin puolella niiden käyttöä voi vähentää kun tiedetään about oikean tyypityksen perusteella mitä tietoa on käytössä missäkin kohtaa. Konfo-uissa toki on vielä vaiheessa tämä tyypittäminen.